### DISCUSSION TITLE
How to automate clicking the "OK" button of the "certificate selection" popup window

### DISCUSSION THREAD

**Discussion by GHRyunosuke** [GCuser99/SeleniumVBA#84] [Q&A] [answered]
While automating some of my daily tasks, I wonder if there is a way to automatically click the "OK" button of the "certificate selection" pop window (in chrome) via SeleniumVBA ?

![image](https://github.com/GCuser99/SeleniumVBA/assets/25561705/d64aa171-5068-473c-b27b-e95feba9fdff)

Similar issue in StackOverflow (still have no idea how to solve this with SeleniumVBA):
https://stackoverflow.com/questions/49512058/selenium-chrome-driver-select-certificate-popup-confirmation-not-working

**Comment by 6DiegoDiego9** [comment]
Try these if you can use that approach:
https://www.perplexity.ai/search/c6b1998c-c355-4804-81c9-c9426ebd0bc1?s=u
https://www.perplexity.ai/search/5033caa0-38bb-44cb-b5d1-e56a3a3ea474?s=u

A more autonomous approach could be to use UIAutomation, identifying the button with Spy++ or the MS LiveInspector:
https://github.com/mhumpher/UIAutomation_VBA

**Reply by GHRyunosuke** [reply]
Thanks a lot for the advices, @6DiegoDiego9.

![image](https://github.com/GCuser99/SeleniumVBA/assets/25561705/eb5a15b9-4ce9-486d-9ec2-997b70f74a84)
I tried to implement above with SeleniumVBA using 
`Set caps = driver.CreateCapabilities`
`caps.SetOption("--policy-file-path", "C:\CProfile\chromePolicy.json")`
`driver.OpenBrowser caps`

However, I got the error of "from invalid argument: unrecognized chrome option: --policy-file-path"

I guess I will just click the OK button myself for now.

**Comment by GCuser99** [comment]
@6DiegoDiego9 - excellent use of generative AI - you are getting to be a pro!

I have sympathy for you @GHRyunosuke - it drives me crazy when I can't automate a task, especially when it's part of a much larger automation flow! :-(

On the running Chrome with a policy file, I don't think it can be set through Selenium capabilities - it has to be set as an argument when starting the chrome.exe binary. One way to test this quickly would be to make a chrome browser shortcut on the desktop and then modify the shortcut properties target field to ```"C:\Program Files\Google\Chrome\Application\chrome.exe --policy-file-path=C:\CProfile\chromePolicy.json"``` and see if that automatically selects the certificate when you manually navigate to the website of interest. If that does work, then we can probably come up with a more programmatic solution. 

Another potential way to test this is to use @6DiegoDiego9's remoteDebugger test procedure to start Chrome up from code and then attach the Selenium WebDriver to it:
```
Sub test_remoteDebugger()
    'this allows one to connect to an existing Edge\Chrome browser instance
    Dim driver As SeleniumVBA.WebDriver
    Dim caps As SeleniumVBA.WebCapabilities
    Dim wshell As Object
    
    Set driver = SeleniumVBA.New_WebDriver
    
    'must start the browser on the debugging port
    Set wshell = CreateObject("WScript.Shell")
    
    'first must kill edge\chrome processes
    wshell.Run "taskkill /f /t /im chrome.exe", 0, True
    
    'Start Chrome browser with a a user-created policy file
    wshell.Run "chrome.exe --remote-debugging-port=9222 --policy-file-path=C:\CProfile\chromePolicy.json"

    driver.StartChrome
    
    Set caps = driver.CreateCapabilities(initializeFromSettingsFile:=False)
    
    'set debugger address to same port as browser
    caps.SetDebuggerAddress "localhost:9222"
    'Debug.Print caps.ToJson
    
    driver.OpenBrowser caps
    
    driver.NavigateTo "https://to_the_annoying_website/"

    driver.Wait 1500
    
    driver.CloseBrowser
    driver.Shutdown
End Sub
```
Again if that works then we can probably figure a way to do this more elegantly....

On the other solution-type which tries to automatically click the OK button - it seems to me that this comment is the key:

https://github.com/SeleniumHQ/selenium/issues/5408#issuecomment-411763750. 

The Certificate Selection dialog blocks execution flow so preventing any means to programmatically send instruction to the dialog from VBA (like VBA's SendKeys) - bummer. The answers in @GHRyunosuke's [stackoverflow reference](https://stackoverflow.com/questions/49512058/selenium-chrome-driver-select-certificate-popup-confirmation-not-working) seem to involve asynchronous execution via multi-threading, which is not supported directly in VBA.

**Comment by GCuser99** [comment]
@6DiegoDiego9, I remember you telling me that you are using VBScript for multi-threading... Do you think running a VBScript from VBA, just before the SeleniumVBA navigation to the desired url, that waits a few seconds and then sends return key(s) via WshShell.SendKeys would work? Like the first answer in @GHRyunosuke's [stackoverflow reference](https://stackoverflow.com/questions/49512058/selenium-chrome-driver-select-certificate-popup-confirmation-not-working)?

I would try to test this myself, but I don't know of a way to replicate @GHRyunosuke's test case - I have never had this happen to me....

**Reply by 6DiegoDiego9** [reply]
@GCuser99 good idea, an easy basic solution could be to run a vbs file like this:
```
WScript.Sleep`` 10000 ' Wait for 10 seconds
Set WshShell = WScript.CreateObject("WScript.Shell")
WshShell.SendKeys "{ENTER}" ' Press the Enter key
(not tested)
```

**Comment by GCuser99** [comment, marked as answer]
So @GHRyunosuke, based on @6DiegoDiego9's response above, maybe this will work without doing the policy stuff (I think it assumes there is only one certificate to select in the dialog):
```
Sub test_click_cert_ok()
    Dim driver As WebDriver
    Dim timeDelay As Long
    Dim vbs As String
    
    Set driver = New WebDriver
    
    driver.StartChrome
    driver.OpenBrowser
    
    'set time delay to wait for url navigation and select cert dialog
    timeDelay = 10000 'ms

    'create the vbs script file to run on different thread
    'this waits a given amount of time, and then sends a return key to active window
    vbs = "WScript.Sleep " & timeDelay & " ' Wait for url to load " & vbCrLf
    vbs = vbs & "Set WshShell = WScript.CreateObject(""WScript.Shell"")" & vbCrLf
    vbs = vbs & "WshShell.SendKeys ""{ENTER}"" ' Press the Enter key"
    
    'save the vbs string to user's temp file
    driver.SaveStringToFile vbs, "%TEMP%\certok.vbs"

    'run the script b4 navigation to url
    Shell "wscript """ & driver.ResolvePath("%TEMP%\certok.vbs") & """ , vbNormalFocus"
    
    'Debug.Print driver.ResolvePath("%TEMP%\certok.vbs")
    
    'navigate to the "problem" url
    driver.NavigateTo "http/the/url/with/cert/select"

    'do other stuff here
    
    'clean up
    'driver.DeleteFiles "%TEMP%\certok.vbs"
    
    driver.CloseBrowser
    driver.Shutdown
End Sub
```

**Reply by GHRyunosuke** [reply]
Hi @GCuser99 and @6DiegoDiego9, you guys are genius!

Running a separate vbs with time delay would perfectly work as "another thread" so that after the main thread generates the certification popup window, the vbs then sends the "ENTER" that clicks the "OK" button.

Thank you so much, this eliminates the only thing I could not automate with SeleniumVBA.

By the way, the "driver.ResolvePath" causes a compile error saying this method or data member could not be found. It seems that there's no such method?

![image](https://github.com/GCuser99/SeleniumVBA/assets/25561705/cc624221-ee2b-40ca-b0e0-40ca9f62d44e)

**Comment by GCuser99** [comment]
That's great news @GHRyunosuke! Thanks for reporting the issue.

Which SeleniumVBA solution are you using? ResolvePath was added on the most recent update (v4.2). I can see it in the SeleniumVBA.xlam. Also, I use the DLL in my projects and this is what I see in the Object Browser, so maybe you just need to update to the latest version....?

<img src="https://github.com/GCuser99/SeleniumVBA/assets/26237126/926255e0-b171-4ea2-825b-229c591452e1" width="50%" height="50%">

**Reply by GHRyunosuke** [reply]
@GCuser99 
Oh, my bad.  I am still using v4.1... time to upgrade.

**Comment by GCuser99** [comment]
@GHRyunosuke, as of v4.3, you can try the following:
```vba
Sub test_click_cert_ok()
    Dim driver As New WebDriver
    Dim keys As New WebKeyboard
    
    driver.StartChrome
    driver.OpenBrowser

    driver.SendKeysToOS _
        keyString:=keys.EnterKey, _
        runOnSeparateThread:=True, _
        timeDelayMS:=10000 'experiment with this value
    
    'navigate to the "problem" url
    driver.NavigateTo "http/the/url/with/cert/select"

    'wait 10000 ms and then send keys to active window

    'do other stuff here
    
    driver.CloseBrowser
    driver.Shutdown
End Sub
```
Additionally, in the case where you know the target window title, then the following should work:
```vba
Sub test_send_keys_to_os_window()
    Dim driver As New WebDriver
    Dim keys As New WebKeyboard
    
    driver.StartChrome
    driver.OpenBrowser

    driver.SendKeysToOS _
        keyString:="string of keys to send", _
        windowTitle:="target window title", _
        runOnSeparateThread:=True, _
        maxTimeToWaitMS:=10000
    
    driver.NavigateTo "http/the/url"

    'wait for the target window to show, up to 10000 ms, then send the keys

    'do other stuff here
    
    driver.CloseBrowser
    driver.Shutdown
End Sub
```
For more information, see [browser-keyboard-interaction-with-sendkeys](https://github.com/GCuser99/SeleniumVBA/wiki#browser-keyboard-interaction-with-sendkeys) in the [Wiki](https://github.com/GCuser99/SeleniumVBA/wiki).

**Reply by GHRyunosuke** [reply]
@GCuser99 
Awesome, confirmed the "SendKeysToOS" method works. Now I can get rid of the extra vbs file and have an all-in-one solution. Thanks a lot!
